flax/errors.py

```
@@ -0,0 +1,67 @@
+class FlaxError(Exception):
+  def __init__(self, message):
+    error_page = 'https://flax.readthedocs.io/en/improve-error/flax.errors.html'
+    module_name = self.__class__.__module__
+    class_name = self.__class__.__name__
+    error_msg = f'{message} ({error_page}#{module_name}.{class_name})'
+    super().__init__(error_msg)
+
+class InvalidFilterError(FlaxError):
+  """
+  A filter should be either a boolean, a string or a container object.
+  """
+  pass
+
+class InvalidScopeError(FlaxError):
+  """
+  A temporary Scope is only valid within the context in which it is created::
+
+    with Scope(variables, rngs=rngs).temporary() as root:
+      y = fn(root, *args, **kwargs)
+      # Here root is valid.
+    # Here root is invalid.
+  """
+  pass
+
+class MutableCollectionError(FlaxError):
+  """
+  Collections that are immutable cannot be modified. When you are applying a
+  Module, you should specify which variable collections are mutable::
+
+    class MyModule(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        ...
+        var = self.variable('batch_stats', 'mean', ...)
+        ...
+    
+    vars = MyModule.init(...)
+    ...
+    logits = MyModule.apply(vars, batch)  # This throws an error.
+    logits = MyModule.apply(vars, batch, mutable=['batch_stats'])  # This works.
+  """
+
+
+class ScopeNamingError(FlaxError):
+  """
+  Scope names should be strings and unique within a subscope::
```

I think we should not use the term "scope" in user-facing errors (both class names and more concretely in error docstrings) because 90% of users (my random assessment) will only interact with Modules and have never heard of scopes. No need to force them to learn about that just because they got this error.

---

(If we're raising this error inside scope.py rather than module.py then maybe we should add another check in module.py to raise a better error, or use a vague name like "VariableNameReuseError".)

---

I agree. For now I've named these error to not include "Scope", and added a TODO saying we should raise these errors in module.py and rename them.